### PR TITLE
[CIRCLE-4801] build image

### PR DESCRIPTION
**Pull request by annguy3n**
TODO: update docker file to be more compatible with circle/dex

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -49,3 +92,13 @@ workflows:
       - unit_tests:
           requires:
             - checkout_code
+      - build_binary:
+          requires:
+            - checkout_code
+      - deploy_image:
```
Comment (annguy3n) [inline .circleci/config.yml:92]: 👍

File: .circleci/config.yml [hunk 2]
```diff
@@ -31,15 +29,60 @@ jobs:
...
+    working_directory: /go/src/github.com/coreos/dex
+    steps:
+      - restore_cache:
+          keys:
+            - v1-repo-{{ .Environment.CIRCLE_SHA1 }}
+      - restore_cache:
+          keys:
+            - v1-binary-{{ .Environment.CIRCLE_SHA1 }}
+      - setup_remote_docker
+      - deploy:
```
Comment (annguy3n) [inline .circleci/config.yml]: duplicated name with L73, probably should rename it to `install_docker_client`

File: .circleci/config.yml [hunk 3]
```diff
@@ -31,15 +29,60 @@ jobs:
...
+            make release-binary
+            mkdir -p _output/bin/dex && cp /go/bin/dex _output/bin/dex
+      - save_cache:
+          key: v1-binary-{{ .Environment.CIRCLE_SHA1 }}
+          paths:
+            - /go/src/github.com/coreos/dex/_output/bin
+
+  deploy_image:
+    docker:
+      - image: circleci/golang:1.8
```
Comment (annguy3n) [inline .circleci/config.yml:58]: Idk i feel like we can use a different image so we dont have to install the client and set up remote docker every time, but that's probably for the future.

File: .circleci/config.yml [hunk 4]
```diff
@@ -31,15 +29,60 @@ jobs:
...
+      - setup_remote_docker
+      - deploy:
+          name: Install Docker client
+          command: |
+            set -x
+            sudo curl -fsSLO https://get.docker.com/builds/Linux/x86_64/docker-17.05.0-ce.tgz && sudo tar --strip-components=1 -xvzf docker-17.05.0-ce.tgz -C /usr/local/bin
+      - deploy:
+          name: Deploy dex Docker image
+          command: |
+            SHORT_SHA=$(git rev-parse --short HEAD)
```
Comment (annguy3n) [inline .circleci/config.yml]: mv this into a file

File: scripts/install-archived-dockerize.sh [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/bin/bash -e
```
Comment (annguy3n) [inline scripts/install-archived-dockerize.sh]: this file isn't used

File: .circleci/config.yml [hunk 6]
```diff
@@ -7,13 +7,11 @@ jobs:
     steps:
       - checkout
       - run: make
-      - run:
-          name: save SHA to a file
-          command: echo $CIRCLE_SHA1 > .circle-sha
       - save_cache:
-          key: v1-repo-{{ checksum ".circle-sha" }}
+          key: v1-repo-{{ .Environment.CIRCLE_SHA1 }}
```
Comment (annguy3n) [inline .circleci/config.yml:11]: 👏  Didn't know we can do that

File: .circleci/config.yml (hunk 3 above)
Comment (lewisl9029) [reply .circleci/config.yml:58]: Totally agree, but I'm not sure which image has docker client installed and couldn't find any in the docs. Any ideas?

File: .circleci/config.yml (hunk 2 above)
Comment (lewisl9029) [reply .circleci/config.yml]: `deploy` is actually just a special `run` step that only gets run once even when parallelism is involved: https://circleci.com/docs/2.0/configuration-reference/#deploy

File: .circleci/config.yml (hunk 3 above)
Comment (lewisl9029) [reply .circleci/config.yml:58]: So I tried to use the official `docker` image, and the problem was it didn't have a bunch of circleci dependencies installed like `tar` `gzip` and whatnot, so we needed a separate step to install things anyways. I figured installing the dependencies from a package manager was probably going to take more time than downloading this one docker binary, so I reverted it.

File: .circleci/config.yml (hunk 2 above)
Comment (annguy3n) [reply .circleci/config.yml]: :/ hmm good to know, probably a comment on top would be good

File: .circleci/config.yml (hunk 3 above)
Comment (annguy3n) [reply .circleci/config.yml:58]: 👍

File: .circleci/config.yml [hunk 7]
```diff
@@ -31,15 +29,51 @@ jobs:
...
+            - v1-binary-{{ .Environment.CIRCLE_SHA1 }}
+      - setup_remote_docker
+      - deploy:
+          name: Install Docker client
+          command: |
+            set -x
+            sudo curl -fsSLO https://get.docker.com/builds/Linux/x86_64/docker-17.05.0-ce.tgz && sudo tar --strip-components=1 -xvzf docker-17.05.0-ce.tgz -C /usr/local/bin
+      - deploy:
+          name: Deploy dex Docker image
+          command: /bin/bash -e scripts/circle-deploy
```
Comment (annguy3n) [inline .circleci/config.yml]: do we still need `/bin/bash -e`? since `scripts/circle-deploy` already has the shebang on top that declares that

**Review by annguy3n — COMMENTED** [review]
couple nits/comments lgtm otherwise

great job!

File: .circleci/config.yml (hunk 7 above)
Comment (lewisl9029) [reply .circleci/config.yml]: I thought that wouldn't work due to lack of `+x` permissions on the file, but apparently it does work so 🎉

Comment (lewisl9029) [reply .circleci/config.yml]: Oh haha it actually didn't work, but it passed anyways because I made that deploy job only run on master...

**Pull request merged**